### DISCUSSION TITLE
useSessionStorage correct way to default for page journey

### DISCUSSION THREAD

**Discussion by Yin117** [juliencrn/usehooks-ts#658] [Help]
Hello,

I have/had a very simple setup that on page 1 of a sequential journey I had:
```js
const [domain, setDomain] = useSessionStorage<string>('domain', process.env.DEV_ONLY_DOMAIN ?? '');
```

However when page 1 loads, the value from `process.env.DEV_ONLY_DOMAIN` doesn't get put into session storage, even though the value does from through into the `domain` variable.

I think for some reason, this line is not running correctly in the hook's code:
```js
  react.useEffect(() => {
    setStoredValue(readValue());
  }, [key]);
```

In that when I pass a default value, it does not go into Session Storage.

This has led me to use a `useEffect` approach to initialize the values into session storage:
```js
  // Initialize Session Storage
  useEffect(() => {
    if (!domain && process.env.DEV_ONLY_DOMAIN) {
      setDomain(process.env.DEV_ONLY_DOMAIN);
    }
  }, []);
```

Thoughts and advice welcome.

**Comment by tamish-max** [comment]
What you’re seeing is expected with the current hook behavior.

`initialValue` is used as the **fallback returned by the hook when the key is missing**. It is **not eagerly written into `sessionStorage` on mount**. The key only gets persisted when you call the setter.

So this is why:
- `domain` has the env value
- but `sessionStorage.getItem('domain')` is still empty until `setDomain(...)` runs

Your `useEffect` initializer is the right approach if you want the default value to be physically stored:

```ts
useEffect(() => {
  if (!domain && process.env.DEV_ONLY_DOMAIN) {
    setDomain(process.env.DEV_ONLY_DOMAIN)
  }
}, [domain, setDomain])
```

I’d think of it as “read-through default”, not “write default on first mount”.

**Reply by Yin117** [reply]
Thanks for the detailed response, that makes sense, its an initialValue, not an initialSessionValue 😄

Sensible, I can see the overhead on doing both.
